proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -7107,6 +7110,34 @@ lemma refill_update_valid_sched_misc[wp]:
           (etcbs_of s) (tcb_sts_of s) (tcb_scps_of s) (tcb_faults_of s) (sc_replies_of s)\<rbrace>"
   unfolding refill_update_def by wpsimp
 
+lemma maybe_add_empty_tail_is_active_sc[wp]:
+  "maybe_add_empty_tail sc_ptr \<lbrace>\<lambda>s. Q (is_active_sc scp s)\<rbrace>"
+  unfolding maybe_add_empty_tail_def
+  apply (wpsimp wp: update_sched_context_wp is_round_robin_wp)
+  by (clarsimp simp: vs_all_heap_simps obj_at_def)
+
+lemma maybe_adfsd_empty_tail_active_sc[wp]:
+  "maybe_add_empty_tail sc_ptr \<lbrace>\<lambda>s. Q (pred_map (refill_ready_sc x) (sc_refill_cfgs_of s) scp)\<rbrace>"
+  unfolding maybe_add_empty_tail_def
+  apply (wpsimp wp: update_sched_context_wp is_round_robin_wp)
+  apply (clarsimp simp: vs_all_heap_simps obj_at_def refill_ready_sc_def)
+  by (case_tac "sc_refills sc"; simp)
+
+lemma maybe_add_empty_tail_is_refill_ready[wp]:
+  "maybe_add_empty_tail sc_ptr \<lbrace>\<lambda>s. Q (is_refill_ready scp s)\<rbrace>"
+  by (rule hoare_lift_Pf[where f = cur_time]; wpsimp)
+
+lemma maybe_add_empty_tail_is_refill_sufficient[wp]:
+  "\<lbrace>is_refill_sufficient usage scp
+    and (\<lambda>s. scp = sc_ptr \<longrightarrow>
```

Comment:
Can we have an empty refill list if it's refill sufficient? Maybe I'm not reading this right

Reply:
Unfortunately, refill sufficiency is expressed in terms of the head refill. And because `hd` is a partial function, it is agnostic as to whether and the refills list is empty, so you can express refill sufficiency without knowing if the head exists or not.